grid-client/deployer/network_deployer.go

```diff
@@ -578,109 +278,92 @@ func (d *NetworkDeployer) updateStateFromDeployments(ctx context.Context, znet *
 			if err != nil {
 				return errors.Wrapf(err, "could not get deployment %d data", dl.ContractID)
 			}
-			if dlData.Name == znet.Name && dl.ContractID != 0 {
-				znet.NodeDeploymentID[nodeID] = dl.ContractID
+			if dlData.Name == znet.GetName() && dl.ContractID != 0 {
+				znet.GetNodeDeploymentID()[nodeID] = dl.ContractID
 			}
 		}
 
-		if contractID, ok := znet.NodeDeploymentID[nodeID]; ok && contractID != 0 {
-			d.tfPluginClient.State.Networks.UpdateNetworkSubnets(znet.Name, znet.NodesIPRange)
+		if contractID, ok := znet.GetNodeDeploymentID()[nodeID]; ok && contractID != 0 {
+			d.tfPluginClient.State.Networks.UpdateNetworkSubnets(znet.GetName(), znet.GetNodesIPRange())
 			d.tfPluginClient.State.StoreContractIDs(nodeID, contractID)
 		}
 	}
 
 	if !updateMetadata {
 		return nil
 	}
-	if err := d.ReadNodesConfig(ctx, znet); err != nil {
-		return errors.Wrapf(err, "could not read node's data for network %s", znet.Name)
+
+	deployments, err := d.deployer.GetDeployments(ctx, znet.GetNodeDeploymentID())
+	if err != nil {
+		return errors.Wrap(err, "failed to get deployment objects")
+	}
+
+	if err := znet.ReadNodesConfig(ctx, deployments); err != nil {
+		return errors.Wrapf(err, "could not read node's data for network %s", znet.GetName())
 	}
 
 	return nil
 }
 
-// InvalidateBrokenAttributes removes outdated attrs and deleted contracts
-func (d *NetworkDeployer) InvalidateBrokenAttributes(znet *workloads.ZNet) error {
-	for node, contractID := range znet.NodeDeploymentID {
-		contract, err := d.tfPluginClient.SubstrateConn.GetContract(contractID)
-		if (err == nil && !contract.IsCreated()) || errors.Is(err, substrate.ErrNotFound) {
-			delete(znet.NodeDeploymentID, node)
-			delete(znet.NodesIPRange, node)
-			delete(znet.Keys, node)
-			delete(znet.WGPort, node)
-		} else if err != nil {
-			return errors.Wrapf(err, "could not get node %d contract %d", node, contractID)
+func (d *NetworkDeployer) calcPublicNode(ctx context.Context, znets []workloads.Network, endpoints map[uint32]net.IP) (map[uint32]struct{}, uint32, error) {
+	allNodes := make(map[uint32]struct{})
+	var multiErr error
+
+	for _, znet := range znets {
+		for _, node := range znet.GetNodes() {
+			allNodes[node] = struct{}{}
+		}
+		if znet.GetPublicNodeID() != 0 {
+			allNodes[znet.GetPublicNodeID()] = struct{}{}
 		}
 	}
-	if znet.ExternalIP != nil && !znet.IPRange.Contains(znet.ExternalIP.IP) {
-		znet.ExternalIP = nil
-	}
-	for node, ip := range znet.NodesIPRange {
-		if !znet.IPRange.Contains(ip.IP) {
-			delete(znet.NodesIPRange, node)
+
+	var publicNode uint32
+	var err error
+	for nodeID := range allNodes {
+		// we check first that there is a public node in all nodes in the networks.
+		// that way we can save extra requests to get a new public node
+		// and its used ports and endpoints. this public node also might not be used
+		// as access node if there is no need for it.
+		if endpoints[nodeID] != nil && endpoints[nodeID].To4() != nil {
```

Comment:
if the network already has a public node here, we don't need to loop over all nodes to check if there's a public node right?